imessage/bluebubbles/api.go

```diff
@@ -0,0 +1,1631 @@
+
+func (bb *blueBubbles) GetGroupAvatar(chatID string) (*imessage.Attachment, error) {
+	bb.log.Trace().Str("chatID", chatID).Msg("GetGroupAvatar")
+
+	chatResponse, err := bb.getChatInfo(chatID)
+	if err != nil {
+		bb.log.Error().Err(err).Str("chatID", chatID).Msg("Failed to fetch chat info")
+		return nil, err
+	}
+
+	if chatResponse.Data.Properties == nil ||
+		len(chatResponse.Data.Properties) < 1 {
+		return nil, nil
+	}
+
+	properties := chatResponse.Data.Properties[0]
+
+	if properties.GroupPhotoGuid == nil {
+		return nil, nil
+	}
+
+	attachment, err := bb.downloadAttachment(*properties.GroupPhotoGuid)
+	if err != nil {
+		bb.log.Error().Err(err).Str("chatID", chatID).Msg("Failed to download group avatar")
+		return nil, err
+	}
+
+	return attachment, nil
+}
+
+// These functions all provide "channels" to allow concurrent processing in the bridge
+func (bb *blueBubbles) MessageChan() <-chan *imessage.Message {
+	return bb.messageChan
+}
+
+func (bb *blueBubbles) ReadReceiptChan() <-chan *imessage.ReadReceipt {
+	return bb.receiptChan
+}
+
+func (bb *blueBubbles) TypingNotificationChan() <-chan *imessage.TypingNotification {
+	return bb.typingChan
+}
+
+func (bb *blueBubbles) ChatChan() <-chan *imessage.ChatInfo {
+	return bb.chatChan
+}
+
+func (bb *blueBubbles) ContactChan() <-chan *imessage.Contact {
+	return bb.contactChan
+}
+
+func (bb *blueBubbles) MessageStatusChan() <-chan *imessage.SendMessageStatus {
+	return bb.messageStatusChan
+}
+
+func (bb *blueBubbles) BackfillTaskChan() <-chan *imessage.BackfillTask {
+	return bb.backfillTaskChan
+}
+
+func (bb *blueBubbles) SendMessage(chatID, text string, replyTo string, replyToPart int, richLink *imessage.RichLink, metadata imessage.MessageMetadata) (*imessage.SendResponse, error) {
+	bb.log.Trace().Str("chatID", chatID).Str("text", text).Str("replyTo", replyTo).Int("replyToPart", replyToPart).Any("richLink", richLink).Interface("metadata", metadata).Msg("SendMessage")
+
+	var method string
+	if replyTo != "" {
+		method = "private-api"
+	} else {
+		// we have to use apple-script and send a second message
+		method = "apple-script"
+	}
+
+	request := SendTextRequest{
+		ChatGUID:            chatID,
+		Method:              method,
+		Message:             text,
+		TempGuid:            fmt.Sprintf("temp-%s", RandString(8)),
+		SelectedMessageGuid: replyTo,
+		PartIndex:           replyToPart,
+	}
+
+	var res SendTextResponse
+
+	err := bb.apiPost("/api/v1/message/text", request, &res)
+	if err != nil {
+		bb.log.Error().Any("response", res).Msg("Failure when sending message to BlueBubbles")
+		return nil, err
+	}
+
+	if res.Status != 200 {
+		bb.log.Error().Int64("statusCode", res.Status).Any("response", res).Msg("Failure when sending message to BlueBubbles")
+
+		return nil, errors.New("could not send message")
+	}
+
+	return &imessage.SendResponse{
+		GUID:    res.Data.GUID,
+		Service: res.Data.Handle.Service,
+		Time:    time.Unix(0, res.Data.DateCreated*int64(time.Millisecond)),
```

Comment:
multiplying/dividing by `time.Millisecond` is a legacy method that shouldn't be used anymore, `time.UnixMilli()` exists now

Reply:
Is `res.Data.DateCreated` in microseconds?

We're multiplying it by 1000 (`time.Millisecond`) and `time.Unix` takes nanoseconds. Does that sound right?

If it is then would we want to use `time.UnixMicro(res.Data.DateCreated)`? Let me know if I'm doing this math completely wrong, I'm playing around with it and getting some strange timestamps :)